File: compiler/src/feedback/items/TypeFeedback.swift
import OrderedCollections

func registerTypeFeedback(into context: FeedbackContext) {
    context.register(
        id: "conflicting-types",
        query: Queries.conflictingTypes,
        rank: { _ in .conflicts },
        location: { data in
            var related = OrderedSet(data.nodes)

            if let source = data.source, data.from != source { related.append(data.from) }

            return (primary: data.source ?? data.from, secondary: related)
        },
        showGraph: true,
    ) { writer, data in
        if let source = data.source {
            writer.write("In ")
            writer.write(source)
            writer.write(", ")
        }

        let hasTypeConstraint = data.trace.contains(where: { $0 is TypeConstraint })

        if hasTypeConstraint, data.nodes.count == 1, data.types.count == 2 {
            writer.write(data.from)
            writer.write(" isn't a ")
            writer.write(.constructed(data.types[1]))
            writer.write(".")
            writer.writeBreak()
        } else {
            writer.write(data.from)
            writer.write(" is a ")
            writer.writeList(separator: "or a") { list in
                for type in data.types { list.add { writer.write(.constructed(type)) } }
            }
            writer.write(", but it can only be one of these.")
        }

        let fromSpan = writer.db[data.from, Syntax.self]?.value.span
        let nodes = data.nodes.filter { writer.db[$0, Syntax.self]?.value.span != fromSpan }

        if nodes.count > 1 {
            writer.writeBreak()
            writer.write(data.from)
            writer.write(" must be the same type as ")
            writer.writeList(separator: "and") { list in
                for node in nodes { list.add { writer.write(node) } }
            }
            writer.write("; double-check these.")
        }

        for constraint in data.trace { writer.write(constraint) }

        var parametersByName: [String: [(source: Node, from: Node)]] = [:]
        for node in data.nodes {
            guard let instantiated = writer.db[node, Instantiated.self],
                let parameter = instantiated.from as Node?,
                let syntax = writer.db[parameter, Syntax.self]?.value as? TypeParameterSyntax
            else { continue }

            parametersByName[String(syntax.name), default: []]
                .append((source: instantiated.sourceNode, from: instantiated.from))
        }

        for name in parametersByName.keys.sorted() {
            let entries = parametersByName[name] ?? []
            guard entries.count > 1, let entry = entries.first else { continue }

            writer.writeBreak()
            writer.write("-  ")
            writer.write(entry.from)
            writer.write(" must have the same type everywhere in this use of ")
            writer.write(entry.source)
            writer.write(".")
        }
    }

    context.register(
        id: "incomplete-type",
        query: Queries.incompleteType,
        rank: { _, _ in .unknown },
        location: { node, _ in (primary: node, secondary: []) },
        showGraph: true,
    ) { writer, node, type in
        writer.write("Missing information for the type of ")
        writer.write(node)
        writer.write(".")
        writer.writeBreak()
        writer.write("Wipple determined this code is ")
        writer.write(.constructed(type))
        writer.write(", but it needs some more information for the ")
        writer.write(code: "_")
        writer.write(" placeholders.")
    }

    context.register(
        id: "unknown-type",
        query: { (db, node, body: (Node) -> Void) in Queries.unknownType(db, node) { body(node) } },
        rank: { _ in .unknown },
        location: { node in (primary: node, secondary: []) },
        showGraph: true,
    ) { writer, node in
        writer.write("Could not determine the type of ")
        writer.write(node)
        writer.write(".")
        writer.writeBreak()
        writer.write(
            "Wipple needs to know the type of this code before running it. Try using a function or assigning it to a variable."
        )
    }

    context.register(
        id: "missing-type",
        query: Queries.fact(MissingTypes.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, missing in
        writer.write(node)

        if missing.types.count == 1, let parameter = missing.types.first {
            writer.write(" is missing a type for ")
            writer.write(parameter)
        } else {
            writer.write(" is missing types for ")
            writer.writeList(separator: "and") { list in
                for parameter in missing.types { list.add { writer.write(parameter) } }
            }
        }

        writer.write(".")
        writer.writeBreak()
        writer.write("Try adding another type here, or double-check your parentheses.")
    }

    context.register(
        id: "extra-type",
        query: Queries.fact(ExtraType.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" doesn't match any parameter of this type.")
        writer.writeBreak()
        writer.write("Try removing this type, or double-check your parentheses.")
    }

    context.register(
        id: "conflicting-instances",
        query: Queries.fact(OverlappingInstances.self),
        rank: { _, _ in .bounds },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, overlapping in
        writer.write(node)
        writer.write(" has multiple overlapping instances: ")
        writer.writeList(separator: "and") { list in
            for instance in overlapping.instances { list.add { writer.write(instance) } }
        }
        writer.write(".")
        writer.writeBreak()
        writer.write(
            "Only one of these instances can be defined at a time. Try making your instance more specific."
        )
    }
}

File: compiler/feedback/src/items/type_feedback.rs
use crate::{FeedbackCtx, FeedbackLocation, FeedbackRank};
use std::collections::BTreeSet;
use wipple_core::{facts::Syntax, typecheck::ty::Ty};
use wipple_queries::{conflicting_types, fact, incomplete_type, unknown_type};
use wipple_syntax::{
    checks::instances::OverlappingInstances,
    types::{ExtraType, MissingTypes},
};

pub fn register(ctx: &mut FeedbackCtx<'_>) {
    ctx.feedback("conflicting-types")
        .query(conflicting_types)
        .rank(|data| {
            if data.source.is_some() {
                FeedbackRank::IndirectConflicts
            } else {
                FeedbackRank::DirectConflicts
            }
        })
        .location(|_, data| {
            let mut secondary = data.nodes.iter().copied().collect::<BTreeSet<_>>();

            if data.source.is_some_and(|source| data.from != source) {
                secondary.insert(data.from);
            }

            FeedbackLocation {
                primary: data.source.unwrap_or(data.from),
                secondary,
            }
        })
        .show_graph()
        .display(|db, writer, _, data| {
            if let Some(source) = data.source {
                writer.string("In ");
                writer.node(source);
                writer.string(", ");
            }

            writer.node(data.from);
            writer.string(" is a ");
            writer.list("or a", |list| {
                for ty in &data.tys {
                    let ty = ty.clone();
                    list.add(move |writer| {
                        writer.ty(db, &Ty::Constructed(ty), None, true);
                    });
                }
            });
            writer.string(", but it can only be one of these.");

            let from_span = db
                .get(data.from)
                .map(|Syntax(syntax)| db.ast(syntax).span(db).clone());

            let nodes = data
                .nodes
                .iter()
                .copied()
                .filter(|node| {
                    db.get(*node).map(|Syntax(syntax)| db.ast(syntax).span(db))
                        != from_span.as_ref()
                })
                .collect::<Vec<_>>();

            if nodes.len() > 1 {
                // Only render visible nodes
                let mut seen = BTreeSet::new();
                let nodes = nodes
                    .into_iter()
                    .filter_map(|node| {
                        let Syntax(syntax) = db.get(node)?;
                        seen.insert(db.ast(syntax).span(db)).then_some(node)
                    })
                    .collect::<Vec<_>>();

                if !nodes.is_empty() {
                    writer.line_break();
                    writer.node(data.from);
                    writer.string(" must be the same type as ");
                    writer.list("and", |list| {
                        for node in nodes {
                            if list.filter(db, node) {
                                list.add(move |writer| writer.node(node));
                            }
                        }
                    });
                    writer.string("; double-check these.");
                }
            }

            writer.traces(db, &data.traces);
        })
        .register();

    ctx.feedback("incomplete-type")
        .query(incomplete_type)
        .rank(|_| FeedbackRank::Unknown)
        .location(|_, (node, _)| FeedbackLocation::from(*node))
        .show_graph()
        .display(|db, writer, _, (node, ty)| {
            writer.string("Missing information for the type of ");
            writer.node(*node);
            writer.string(".");
            writer.line_break();
            writer.string("Wipple determined this code is a ");
            writer.ty(db, &Ty::Constructed((*ty).clone()), None, true);
            writer.string(", but it needs some more information for the ");
            writer.code("_");
            writer.string(" placeholders.");
        })
        .register();

    ctx.feedback("unknown-type")
        .query(unknown_type)
        .rank(|_| FeedbackRank::Unknown)
        .show_graph()
        .display(|_db, writer, node, _| {
            writer.string("Could not determine the type of ");
            writer.node(node);
            writer.string(".");
            writer.line_break();
            writer.string(
                "Wipple needs to know the type of this code before running it. Try using a function or assigning it to a variable.",
            );
        })
        .register();

    ctx.feedback("missing-type")
        .query(fact::<MissingTypes>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, MissingTypes(parameters)| {
            writer.node(node);

            if let &[parameter] = parameters.as_slice() {
                writer.string(" is missing a type for ");
                writer.node(parameter);
            } else {
                writer.string(" is missing types for ");
                writer.list("and", |list| {
                    for &parameter in parameters {
                        list.add(move |writer| writer.node(parameter));
                    }
                });
            }

            writer.string(".");
            writer.line_break();
            writer.string("Try adding another type here, or double-check your parentheses.");
        })
        .register();

    ctx.feedback("extra-type")
        .query(fact::<ExtraType>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" doesn't match any parameter of this type.");
            writer.line_break();
            writer.string("Try removing this type, or double-check your parentheses.");
        })
        .register();

    ctx.feedback("conflicting-instances")
        .query(fact::<OverlappingInstances>)
        .rank(|_| FeedbackRank::Bounds)
        .display(|_db, writer, node, OverlappingInstances(instances)| {
            writer.node(node);
            writer.string(" has multiple overlapping instances: ");
            writer.list("and", |list| {
                for &instance in instances {
                    list.add(move |writer| writer.node(instance));
                }
            });
            writer.string(".");
            writer.line_break();
            writer.string(
                "Only one of these instances can be defined at a time. Try making your instance more specific.",
            );
        })
        .register();
}
